posts/2021/11/pydata-extensibility-vision.md

```diff
@@ -0,0 +1,191 @@
+For a demo, let's begin with a tool from the Python standard library: `functools.singledispatch`.
+```python
+from functools import singledispatch
+import scipy.ndimage as _scipy_ndimage
+
+@singledispatch
+def laplace(input, output=None, mode="reflect", cval=0.0):
+    return _scipy_ndimage.laplace(input, output=output, mode=mode, cval=cval)
+
+# export `laplace` as `scipy.ndimage.laplace`
+
+# In CuPy codebase
+# from scipy.ndimage import laplace
+import cupyx.scipy.ndimage as _cupyx_ndimage
+
+@laplace.register(cupy.ndarray)
+def cupy_laplace(input, output=None, mode="reflect", cval=0.0):
+    return _cupyx_ndimage.laplace(input, output=output, mode=mode, cval=cval)
+
+# run registration code when CuPy is imported
+```
+
+We begin by defining the base `laplace` function, it is a fallback implementation that is going to be called for non-registered types of `input`. Next `laplace.register` is used to connect the implementation that is specific to `cupy.ndarray` with the base `laplace` function. We get a function in the `scipy.ndimage` module that is now compatible with CuPy arrays. Similarly, it can be extended to work with Dask arrays.
+```python
+from scipy.ndimage import laplace
+import numpy as np
+import cupy as cp
+
+numpy_img = np.random.normal(size=(256, 256))
+cupy_img = cp.array(numpy_img)
+
+print(type(laplace(numpy_img)))
+# <class 'numpy.ndarray'>
+
+print(type(laplace(cupy_img)))
+# <class 'cupy._core.core.ndarray'>
+```
+
+Since Python doesn't provide standard tools for using multiple dispatch there are several dedicated libraries:
+* _multimethod_ - pure Python implementation of multiple dispatch with caching of argument types,
+* _plum-dispatch_ - implementation of multiple dispatch that follows the ideas from Julia,
+* _uarray_ is a generic backend/multiple dispatch library developed to provide NumPy-independent but similar to NumPy's dispatch functionality.
+
+First two options are very similar in functionality and usability. But the last option, in addition to multiple dispatch, offers granular control using context managers and provides a way to switch backends for the same array type. It's also possible to make the `uarray` dispatcher work without a context manager making the mechanism implicit and similar to other dispatch libraries. Whether we should encourage the implicit registration of backends or not is an open design issue and there's previous discussion on the topic in [SciPy #14266](https://github.com/scipy/scipy/issues/14266). Hameer Abbasi, one of the `uarray` authors, wrote [a blog post](https://labs.quansight.org/blog/2019/07/uarray-update-api-changes-overhead-and-comparison-to-__array_function__/) about the motivation for the `uarray` and how it compares to NumPy's `__array_function__` dispatch protocol.
```

Comment:
I'll try and tweak this a little I think, since it's been a topic of a lot of messy discussions. Let me try and state why `uarray` is preferred., e.g.:

1. we will have modules where multiple other libraries will want to provide alternatives for the same array type (true for `fft` and `linalg` clearly, perhaps more)
2. we don't want two different systems for the same job (and given 1 above, uarray is the one good candidate)
3. (add simplest example of another feature specific to uarray)

@thomasjpfan may also have some thoughts here, since he is fairly new to uarray and immediately liked it when he wrapped his head around it.

Reply:
My preference for `uarray` is very pragmatic:

1. SciPy already ships `fft` that integrates with `uarray` and there is momentum for other submodules.
2. Covers use cases that `__array_function__` or `__array_namespace__` does not.

In terms of comparing it to `multimethod` or `plum-dispatch`, I think point 1 already gives `uarray` an edge. Furthermore, `uarray` defines protocols to communicate between API author and backend providers.